File: backend/src/utils/generateMockData.ts
import fs from 'fs';
import path from 'path';

const classes = ['peaton', 'motocicleta', 'automovil', 'bus_transcaribe', 'ciclista'] as const;

function rand(min: number, max: number) {
  return Math.random() * (max - min) + min;
}

function randomDetection(trackId: string) {
  const className = classes[Math.floor(Math.random() * classes.length)];
  return {
    track_id: trackId,
    class_name: className,
    confidence: Number(rand(0.72, 0.98).toFixed(2)),
    bbox: {
      x: Number(rand(100, 1200).toFixed(1)),
      y: Number(rand(120, 620).toFixed(1)),
      width: Number(rand(40, 160).toFixed(1)),
      height: Number(rand(60, 220).toFixed(1))
    }
  };
}

function generateFrames(totalFrames = 80) {
  const now = Date.now();
  const frames = [];
  for (let i = 0; i < totalFrames; i++) {
    const detCount = Math.floor(rand(4, 12));
    const detections = Array.from({ length: detCount }, (_, d) => randomDetection(`${d + 1}`));

    frames.push({
      camera_id: 'cam-001-cartagena-centro',
      timestamp: new Date(now + i * 200).toISOString(),
      gps: { lat: 10.4236, lng: -75.5457 },
      frame_size: { width: 1280, height: 720 },
      crosswalk_polygon: [
        { x: 520, y: 380 },
        { x: 880, y: 380 },
        { x: 980, y: 580 },
        { x: 470, y: 580 }
      ],
      detections
    });
  }
  return frames;
}

const outputPath = path.resolve(process.cwd(), '../data/sample-sensecraft-json.json');
const data = generateFrames();
fs.writeFileSync(outputPath, JSON.stringify(data, null, 2), 'utf-8');
console.log(`Mock generado en ${outputPath} con ${data.length} frames.`);

File: backend_py/tools/generate_mock_data.py
from __future__ import annotations

import json
from datetime import datetime, timedelta, timezone
from pathlib import Path
import random

ROOT = Path(__file__).resolve().parents[2]
OUT_PATH = ROOT / "data" / "sample-sensecraft-json.json"


def make_frame(ts: datetime, idx: int) -> dict:
    base_x = 520 + (idx % 12) * 10
    return {
        "camera_id": "cam-001-cartagena-centro",
        "timestamp": ts.isoformat().replace("+00:00", "Z"),
        "gps": {"lat": 10.4236, "lng": -75.5457},
        "frame_size": {"width": 1280, "height": 720},
        "crosswalk_polygon": [
            {"x": 520, "y": 380},
            {"x": 880, "y": 380},
            {"x": 980, "y": 580},
            {"x": 470, "y": 580},
        ],
        "detections": [
            {
                "track_id": f"p-{100 + idx}",
                "class_name": "peaton",
                "confidence": 0.88,
                "bbox": {
                    "x": float(base_x),
                    "y": float(430 + random.randint(-8, 8)),
                    "width": 60.0,
                    "height": 150.0,
                },
            },
            {
                "track_id": f"a-{200 + idx}",
                "class_name": "automovil",
                "confidence": 0.9,
                "bbox": {
                    "x": float(560 + idx * 6),
                    "y": float(455 + random.randint(-6, 6)),
                    "width": 120.0,
                    "height": 90.0,
                },
            },
        ],
    }


def main() -> None:
    OUT_PATH.parent.mkdir(parents=True, exist_ok=True)
    start = datetime.now(timezone.utc)
    frames = [make_frame(start + timedelta(milliseconds=240 * i), i) for i in range(60)]
    OUT_PATH.write_text(json.dumps(frames, ensure_ascii=False, indent=2), encoding="utf-8")
    print(f"Archivo generado: {OUT_PATH}")


if __name__ == "__main__":
    main()
